im system

genius-child-alice opened the pull request:
本次 PR 新增了IM系统功能，包括消息发送、会话管理、消息已读状态、消息撤回等功能。cc @cfdxkk 为了阻止blocked角色使用im功能，应用了rbac角色检查，给user角色添加以下路径权限就行了：/im/sendMessage、/im/conversationList、/im/messageList、/im/markMessageRead、/im/conversation、/im/message、/im/unreadCount、/im/recallMessage。

## ✨ 新增功能

### 1. 发送消息
- **端点**: `POST /im/sendMessage`
- **功能**: 向指定用户发送消息
- **请求参数**:
  - `receiverUid`: 接收者 UID
  - `messageType`: 消息类型（text/image/file，目前仅支持 text）
  - `content`: 消息内容
- **业务规则**:
  - 未关注用户需等待对方回复后才能继续发送（防骚扰机制）
  - 不能给自己发消息
  - 被拉黑的用户无法发送消息

### 2. 获取会话列表
- **端点**: `GET /im/conversationList`
- **功能**: 获取当前用户的会话列表
- **查询参数**:
  - `page`: 页码（可选，默认 1）
  - `pageSize`: 每页数量（可选，默认 20）
- **返回**: 会话列表，包含对方用户信息、最后一条消息、未读消息数等

### 3. 获取消息列表
- **端点**: `GET /im/messageList`
- **功能**: 获取指定会话的消息列表
- **查询参数**:
  - `conversationId`: 会话 ID（必需）
  - `page`: 页码（可选，默认 1）
  - `pageSize`: 每页数量（可选，默认 50）
  - `markAsRead`: 是否在获取后自动标记为已读（可选，默认 false）

### 4. 标记消息已读
- **端点**: `POST /im/markMessageRead`
- **功能**: 标记消息为已读状态
- **请求参数**:
  - `conversationId`: 会话 ID（必需）
  - `messageIds`: 消息 ID 列表（可选，为空则标记该会话所有未读消息）

### 5. 删除会话
- **端点**: `DELETE /im/conversation`
- **功能**: 软删除会话（仅对当前用户隐藏，不影响对方）
- **请求参数**:
  - `conversationId`: 会话 ID

### 6. 删除消息
- **端点**: `DELETE /im/message`
- **功能**: 软删除消息（仅对当前用户隐藏，不影响对方）
- **请求参数**:
  - `messageId`: 消息 ID

### 7. 获取未读消息总数
- **端点**: `GET /im/unreadCount`
- **功能**: 获取当前用户所有会话的未读消息总数

### 8. 撤回消息
- **端点**: `POST /im/recallMessage`
- **功能**: 撤回已发送的消息（无时间限制，可随时撤回）
- **请求参数**:
  - `messageId`: 消息 ID
- **说明**: 撤回后的消息会显示为 "[消息已撤回]"，但原始内容仍保留在数据库中

## 🗄️ 数据库结构变更

### 新增集合

#### 1. `im-conversation`（IM会话表）
存储两个用户之间的会话信息。

**字段说明**:
- `conversationId`: 会话 ID（唯一，格式：`conv_{uuid1}_{uuid2}`，按字典序排序）
- `user1Uuid`: 用户1的 UUID
- `user2Uuid`: 用户2的 UUID
- `lastMessageId`: 最后一条消息的 ID
- `lastMessageTime`: 最后一条消息的时间
- `user1UnreadCount`: 用户1的未读消息数
- `user2UnreadCount`: 用户2的未读消息数
- `user1Deleted`: 用户1是否已删除会话
- `user1DeletedTime`: 用户1删除会话的时间
- `user2Deleted`: 用户2是否已删除会话
- `user2DeletedTime`: 用户2删除会话的时间
- `createDateTime`: 创建时间
- `editDateTime`: 最后编辑时间

**索引**:
- 复合唯一索引：`{ user1Uuid: 1, user2Uuid: 1 }` - 确保两个用户之间只有一个会话
- 索引：`{ user1Uuid: 1, editDateTime: -1 }` - 用于查询用户1的会话列表
- 索引：`{ user2Uuid: 1, editDateTime: -1 }` - 用于查询用户2的会话列表

#### 2. `im-message`（IM消息表）
存储具体的消息内容。

**字段说明**:
- `messageId`: 消息 ID（唯一，格式：`msg_{uuid}`）
- `conversationId`: 会话 ID
- `senderUuid`: 发送者 UUID
- `receiverUuid`: 接收者 UUID
- `messageType`: 消息类型（text/image/file）
- `content`: 消息内容
- `isRead`: 是否已读
- `readTime`: 已读时间
- `senderDeleted`: 发送者是否已删除
- `receiverDeleted`: 接收者是否已删除
- `isRecalled`: 是否已撤回
- `recalledTime`: 撤回时间
- `createDateTime`: 创建时间
- `editDateTime`: 最后编辑时间

**索引**:
- 索引：`{ conversationId: 1, createDateTime: -1 }` - 用于查询会话的消息列表
- 索引：`{ receiverUuid: 1, isRead: 1, createDateTime: -1 }` - 用于查询用户的未读消息

### 数据库兼容性

✅ **未修改任何原有数据库结构**
- 所有变更均为新增集合
- 不影响现有功能
- 无需数据迁移

## 🔒 安全特性

1. **防骚扰机制**: 未关注用户需等待对方回复后才能继续发送消息
2. **拉黑保护**: 被拉黑的用户无法发送消息
3. **权限验证**: 所有操作均需验证用户身份和权限
4. **软删除**: 消息和会话采用软删除机制，数据不会真正丢失

## 📝 技术实现

- **事务支持**: 关键操作（发送消息、标记已读）使用 MongoDB 事务保证数据一致性
- **索引优化**: 针对常用查询场景建立了合适的索引
- **聚合查询**: 会话列表使用 MongoDB 聚合管道优化查询性能
- **软删除**: 消息和会话支持软删除，支持恢复

## 🧪 测试建议

- [ ] 测试消息发送功能（包括防骚扰机制）
- [ ] 测试会话列表获取和排序
- [ ] 测试消息列表获取和分页
- [ ] 测试消息已读状态更新
- [ ] 测试消息和会话的软删除
- [ ] 测试消息撤回功能
- [ ] 测试未读消息计数
- [ ] 测试并发场景下的数据一致性

## 📚 相关文件

- `src/dbPool/schema/ImSchema.ts` - IM 数据库 Schema 定义
- `src/service/ImService.ts` - IM 业务逻辑服务
- `src/controller/ImController.ts` - IM API 控制器
- `src/controller/ImControllerDto.ts` - IM API DTO 定义
- `src/route/router.ts` - 路由配置（已添加 IM 相关路由）

genius-child-alice committed:
feat: add im system implementation

cfdxkk commented:
你有没有研究过 k8s 容器集群中的 websocket，轮询还是不太好，我想要 websocket 实现

genius-child-alice commented:
> 你有没有研究过 k8s 容器集群中的 websocket，轮询还是不太好，我想要 websocket 实现

我可以试试@cfdxkk

genius-child-alice committed:
feat: 添加日志服务并替换多处console输出为logging调用

refactor(i18n): 标准化客户端语言标识符处理逻辑

feat(dbPool): 在UpdateType中添加$inc操作符支持数字自增

docs(locales): 更新多语言邮件模板内容及键名

style: 统一代码风格并移除不必要的注释

fix: 修复多处潜在错误处理逻辑

chore: 更新依赖及配置

genius-child-alice committed:
refactor(ImService): 移除消息撤回时间限制检查

移除消息撤回服务中的2分钟时间限制检查逻辑，简化撤回功能

genius-child-alice committed:
feat(controller): 为IM控制器添加RBAC权限验证

在发送消息、获取会话列表等所有控制器方法中添加RBAC权限检查，确保只有授权用户才能访问相关接口

YasaitoriOvO commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,207 @@
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
+	
+	const data = ctx.request.body as Partial<SendMessageRequestDto>
+	
+	const sendMessageRequest: SendMessageRequestDto = {
+		receiverUid: data.receiverUid ?? -1,
+		messageType: data.messageType ?? IM_MESSAGE_TYPE.text,
+		content: data.content ?? '',
+	}
+	
+	const sendMessageResult = await sendMessageService(sendMessageRequest, uuid, token)
+	ctx.body = sendMessageResult
+	await next()
+}
+
+/**
+ * 获取会话列表
+ */
+export const getConversationListController = async (ctx: koaCtx, next: koaNext) => {
+	const uuid = ctx.cookies.get('uuid')
+	const token = ctx.cookies.get('token')
+	
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
+	
+	const getConversationListRequest: GetConversationListRequestDto = {
+		page: ctx.query.page ? parseInt(ctx.query.page as string) : undefined,
+		pageSize: ctx.query.pageSize ? parseInt(ctx.query.pageSize as string) : undefined,
+	}
+	
+	const getConversationListResult = await getConversationListService(getConversationListRequest, uuid, token)
+	ctx.body = getConversationListResult
+	await next()
+}
```
```suggestion
/**
 * 获取会话列表
 */
export const getConversationListController = async (ctx: koaCtx, next: koaNext) => {
	const uuid = ctx.cookies.get('uuid')
	const token = ctx.cookies.get('token')
	const page = ctx.query.page as string
	const pageSize = ctx.query.pageSize as string
	
	// RBAC 权限验证
	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
		return
	}
	
	const getConversationListRequest: GetConversationListRequestDto = {
		pagination: {
			page: parseInteger(page || '1') ?? 1,
			pageSize: Number.isFinite(parseInteger(pageSize)) ? parseInteger(pageSize) : Number.MAX_SAFE_INTEGER,
  },
	
	const getConversationListResult = await getConversationListService(getConversationListRequest, uuid, token)
	ctx.body = getConversationListResult
	await next()
}
```

```page``` 与 ```pageSize``` 建议包裹在 ```pagination``` 下
同时尽量使用 ```parseInteger``` 替换 ```parseInt``` 详见 #53

```typescript
// 此为伪代码
const getSomethingRequest = {
	pagination: {
		page: parseInteger(page || '1') ?? 1,
		pageSize: Number.isFinite(parseInteger(pageSize)) ? parseInteger(pageSize) : Number.MAX_SAFE_INTEGER,
  },
}
```

cfdxkk replied:
@genius-child-alice 
如果解决了可以点 ↓ 下面的按钮标记为解决。

YasaitoriOvO commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,207 @@
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
+	
+	const getConversationListRequest: GetConversationListRequestDto = {
+		page: ctx.query.page ? parseInt(ctx.query.page as string) : undefined,
+		pageSize: ctx.query.pageSize ? parseInt(ctx.query.pageSize as string) : undefined,
+	}
+	
+	const getConversationListResult = await getConversationListService(getConversationListRequest, uuid, token)
+	ctx.body = getConversationListResult
+	await next()
+}
+
+/**
+ * 获取消息列表
+ */
+export const getMessageListController = async (ctx: koaCtx, next: koaNext) => {
+	const uuid = ctx.cookies.get('uuid')
+	const token = ctx.cookies.get('token')
+	
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
+	
+	const getMessageListRequest: GetMessageListRequestDto = {
+		conversationId: ctx.query.conversationId as string,
+		page: ctx.query.page ? parseInt(ctx.query.page as string) : undefined,
+		pageSize: ctx.query.pageSize ? parseInt(ctx.query.pageSize as string) : undefined,
+		markAsRead: ctx.query.markAsRead === 'true',
+	}
+	
+	const getMessageListResult = await getMessageListService(getMessageListRequest, uuid, token)
+	ctx.body = getMessageListResult
+	await next()
+}
```
```suggestion
/**
 * 获取消息列表
 */
export const getMessageListController = async (ctx: koaCtx, next: koaNext) => {
	const uuid = ctx.cookies.get('uuid')
	const token = ctx.cookies.get('token')
  const page = ctx.query.page as string  
  const pageSize = ctx.query.pageSize as string  
	
	// RBAC 权限验证
	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
		return
	}
	
	const getMessageListRequest: GetMessageListRequestDto = {
		conversationId: ctx.query.conversationId as string,
    markAsRead: ctx.query.markAsRead === 'true',
    pagination: {
			page: parseInteger(page || '1') ?? 1,
			pageSize: Number.isFinite(parseInteger(pageSize)) ? parseInteger(pageSize) : Number.MAX_SAFE_INTEGER,
		},
	},
	
	const getMessageListResult = await getMessageListService(getMessageListRequest, uuid, token)
	ctx.body = getMessageListResult
	await next()
}
```
理由同上

YasaitoriOvO commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,207 @@
+	
+	const getUnreadMessageCountResult = await getUnreadMessageCountService(getUnreadMessageCountRequest, uuid, token)
+	ctx.body = getUnreadMessageCountResult
+	await next()
```
```suggestion
	const getUnreadMessageCountResult = await getUnreadMessageCountService(uuid, token)
	ctx.body = getUnreadMessageCountResult
	await next()
```

YasaitoriOvO commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,207 @@
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
```
```suggestion
```

YasaitoriOvO commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,207 @@
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
```
```suggestion
```

YasaitoriOvO commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,207 @@
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
+		return
+	}
```
```suggestion
```

YasaitoriOvO committed:
🎨 QoL: im system

genius-child-alice committed:
refactor(ImService): 简化消息已读标记的用户判断逻辑

移除不必要的数组排序操作，直接比较当前用户与会话中的user1Uuid

YasaitoriOvO commented on src/dbPool/schema/ImSchema.ts:
```diff
@@ -0,0 +1,113 @@
+		/** 发送者是否已删除 - 非空 */
+		senderDeleted: { type: Boolean, required: true, default: false },
+		/** 接收者是否已删除 - 非空 */
+		receiverDeleted: { type: Boolean, required: true, default: false },
+		/** 是否已撤回 - 非空 */
+		isRecalled: { type: Boolean, required: true, default: false },
+		/** 撤回时间 */
```
@cfdxkk

你觉得撤回的消息有没有必要再单独开一个 ```collection``` 存放？

cfdxkk replied:
不用了，如果拷贝到另一个表，查询会多一倍。

YasaitoriOvO commented on src/route/router.ts:
```diff
@@ -867,6 +840,58 @@ router.get('/secret/getStgEnvBackEndSecret', getStgEnvBackEndSecretController) /
+router.get('/im/conversationList', getConversationListController) // 获取会话列表
+// https://localhost:9999/im/conversationList?page=1&pageSize=20
+// cookie: uuid, token
+
+router.get('/im/messageList', getMessageListController) // 获取消息列表
+// https://localhost:9999/im/messageList?conversationId=conv_xxx_yyy&page=1&pageSize=50&markAsRead=true
+// cookie: uuid, token
+
+router.post('/im/markMessageRead', markMessageReadController) // 标记消息已读
+// https://localhost:9999/im/markMessageRead
+// cookie: uuid, token
+// {
+// 	"conversationId": "conv_xxx_yyy",
+// 	"messageIds": ["msg_xxx", "msg_yyy"] // 可选，为空则标记该会话所有未读消息为已读
+// }
+
+router.delete('/im/conversation', deleteConversationController) // 删除会话
+// https://localhost:9999/im/conversation
+// cookie: uuid, token
+// {
+// 	"conversationId": "conv_xxx_yyy"
+// }
+
+router.delete('/im/message', deleteMessageController) // 删除消息
+// https://localhost:9999/im/message
+// cookie: uuid, token
+// {
+// 	"messageId": "msg_xxx"
+// }
+
+router.get('/im/unreadCount', getUnreadMessageCountController) // 获取未读消息总数
+// https://localhost:9999/im/unreadCount
+// cookie: uuid, token
+
+router.post('/im/recallMessage', recallMessageController) // 撤回消息
+// https://localhost:9999/im/recallMessage
+// cookie: uuid, token
+// {
+// 	"messageId": "msg_xxx"
+// }
```
这里可以往上移动 用于分类的空格也可以多放点

genius-child-alice committed:
🎨重构分页参数结构为嵌套对象

genius-child-alice committed:
🎨修正ImControllerDto的注释

cfdxkk commented on src/controller/ImControllerDto.ts:
```diff
@@ -0,0 +1,204 @@
+	success: boolean
+	message: string
+	messageId?: string
+	conversationId?: string
```
Dto 文件会被复制到前端，需要每条类型都添加注释。
该文件及其它 Dto 文件同理。

cfdxkk commented on src/controller/ImController.ts:
```diff
@@ -0,0 +1,224 @@
+	const token = ctx.cookies.get('token')
+
+	// RBAC 权限验证
+	if (!await isPassRbacCheck({ uuid, apiPath: ctx.path }, ctx)) {
```
获取未读消息应该不必 RBAC，即使被封禁的用户，这种基础功能应该也能用。
另外，这个接口可能会被频繁调用，不在 RBAC 下可以增加性能。

cfdxkk commented on src/dbPool/schema/ImSchema.ts:
```diff
@@ -0,0 +1,113 @@
+		/** 系统专用字段-创建时间 - 非空 */
+		createDateTime: { type: Number, required: true },
+		/** 系统专用字段-最后编辑时间 - 非空 */
+		editDateTime: { type: Number, required: true },
```
```suggestion
		/** 系统专用字段-创建时间 - 非空 */
		createdDateTime: { type: Number, required: true },
		/** 系统专用字段-创建者 - 非空 */
		createdBy: { type: String, required: true },
		/** 系统专用字段-最后编辑时间 - 非空 */
		editedDateTime: { type: Number, required: true },
		/** 系统专用字段-最后编辑者 - 非空 */
		editedBy: { type: String, required: true },
```
逐渐迭代到四个系统用字段，新增的集合都加上吧

cfdxkk commented on src/dbPool/schema/ImSchema.ts:
```diff
@@ -0,0 +1,113 @@
+		/** 系统专用字段-创建时间 - 非空 */
+		createDateTime: { type: Number, required: true },
+		/** 系统专用字段-最后编辑时间 - 非空 */
+		editDateTime: { type: Number, required: true },
```
```suggestion
		/** 系统专用字段-创建时间 - 非空 */
		createdDateTime: { type: Number, required: true },
		/** 系统专用字段-创建者 - 非空 */
		createdBy: { type: String, required: true },
		/** 系统专用字段-最后编辑时间 - 非空 */
		editedDateTime: { type: Number, required: true },
		/** 系统专用字段-最后编辑者 - 非空 */
		editedBy: { type: String, required: true },
```
同上，逐渐迭代到四个系统用字段，新增的集合都加上

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+/**
+ * 生成会话ID（确保两个用户之间的会话ID唯一且一致）
+ */
+const generateConversationId = (user1Uuid: string, user2Uuid: string): string => {
```
UUID 不是公开的。这个 conversationId 会暴露对方 UUID，改成 UID (number)

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+		logging('ERROR', '检查用户是否被拉黑失败：', error)
+		return false
+	}
+}
```
```suggestion
```

请使用 BlockService.ts 中的函数
checkBlockUserService - 用于检查对方是否已经被你屏蔽或隐藏
checkIsBlockedByOtherUserService - 用于检查自己是否被对方屏蔽

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+		logging('ERROR', '检查用户是否关注失败：', error)
+		return false
+	}
+}
```
可选：完善该函数并移动到 FeedService.ts 中以改进关注/动态相关功能。

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+
+/**
+ * 检查是否已经发送过消息但对方未回复
+ */
```
JSDoc 中添加变量说明

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+/**
+ * 获取或创建会话
+ */
+const getOrCreateConversation = async (user1Uuid: string, user2Uuid: string): Promise<{ success: boolean; conversation?: InferSchemaType<typeof ImConversationSchema.schemaInstance> }> => {
```
实际上是获取不到对方的 UUID 的，只能获取到 UID
UUID 不应暴露给除自己之外的其他用户，只在后端程序内部查询使用

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+				return { success: false, message: '发送消息失败：对方未回复你的上一条消息，且你未关注对方' }
+			}
+		}
+
```
现在的逻辑是如果 A 关注了 B，那么 A 就可以无限发消息给 B，不是应该反过来，如果 B 关注了 A，那么 A 就可以无限发消息给 B，否则只能在 B 回复 A 之前发送一条消息？

另外一条消息的设定有点恶心，很容易话没说完就不小心发出去了然后再发不了了。我认为可以增加到 3 条，以及在用户发送消息前就显示提醒剩余可发送条数（或者提醒仅能发送一条）。

cc @Aira-Sakuranomiya @otomad

Aira-Sakuranomiya replied:
我觉得不应该限定能发送多少条，比如说第一次我说了某件事情，对方懒得回。但是第二次我又想说一些很重要的东西结果发现没法说了，这就很恶心了。

应该让用户设置私信权限：

- 允许所有人给自己发送私信
- 只允许自己关注的人发送私信给自己
- 不允许任何人给自己发送私信

还有就是陌生人给自己发送的私信可以折叠到一个列表里。

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+				logging('ERROR', '发送消息失败：消息内容过长')
+				return { success: false, message: '发送消息失败：消息内容过长' }
+			}
+		}
```
```suggestion
		}
		
		// TODO: 增加消息审核 / 关键词过滤
```

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+
+		// 获取接收者UUID
+		const receiverUuid = await getUserUuid(receiverUid)
+		if (!receiverUuid) {
+			logging('ERROR', '发送消息失败：接收者不存在')
+			return { success: false, message: '发送消息失败：接收者不存在' }
+		}
+
+		// 不能给自己发消息
+		if (senderUuid === receiverUuid) {
+			logging('ERROR', '发送消息失败：不能给自己发消息')
+			return { success: false, message: '发送消息失败：不能给自己发消息' }
+		}
```
🤔 大多数以 IM 为核心功能的项目都支持给自己发消息，但大多数 IM 非核心功能的项目都不支持。
我们去掉这个限制会有什么副作用吗？

genius-child-alice replied:
没有什么副作用，相当于跨设备发信息，但是要考虑未读消息数、关注检查、消息列表显示等逻辑@cfdxkk

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+
+		try {
+			// 获取或创建会话
+			const conversationResult = await getOrCreateConversation(senderUuid, receiverUuid)
```
session 也传递进去吧

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+			senderUuid: item.senderUuid || '',
+			receiverUuid: item.receiverUuid || '',
+			messageType: item.messageType || IM_MESSAGE_TYPE.text,
+			content: (item.isRecalled ? '[消息已撤回]' : item.content) || '', // 如果已撤回，显示"[消息已撤回]"
```
```suggestion
			content: (item.isRecalled ? '' : item.content) || '',
```

加了这个反而容易误导，反正有 `isRecalled` 来判定是否撤回，直接返回空完事。

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1206 @@
+				$match: {
+					$or: [
+						{ user1Uuid: uuid, user1Deleted: { $ne: true } },
+						{ user2Uuid: uuid, user2Deleted: { $ne: true } },
```
```suggestion
						{ user1Uuid: uuid, user1Deleted: false },
						{ user2Uuid: uuid, user2Deleted: false },
```

参考 Schema 中对集合中字段的定义
``` typescript
...
/** 用户1是否已删除会话 - 非空 */
user1Deleted: { type: Boolean, required: true, default: false },
...
/** 用户2是否已删除会话 - 非空 */
user2Deleted: { type: Boolean, required: true, default: false },
...
```
类型为 Boolean 而且 required: true，我想 `user1Deleted: { $ne: true }` 应该是等价于 `user1Deleted: false`
`user1Deleted: { $ne: true }` 的性能更差且难以索引，请检查并修改本文件中其他类似问题。

genius-child-alice committed:
Merge remote-tracking branch 'origin/develop' into genius-child-alice/feature/im-system

genius-child-alice committed:
Merge remote-tracking branch 'origin/develop' into genius-child-alice/feature/im-system

genius-child-alice committed:
Merge branch 'genius-child-alice/feature/im-system' of https://github.com/KIRAKIRA-DOUGA/KIRAKIRA-Rosales into genius-child-alice/feature/im-system

cfdxkk commented:
重点：UUID 不是公开的，创建聊天的时候获取不到对方的 UUID

genius-child-alice committed:
🎨 为IM控制器DTO添加详细的字段注释

为所有IM控制器相关的DTO类型添加详细的JSDoc注释，明确每个字段的含义和用途，提高代码可读性和维护性

genius-child-alice committed:
🐛 移除IM控制器中的RBAC权限检查

删除deleteConversationController、deleteMessageController和getUnreadMessageCountController中的RBAC权限验证代码，简化控制器逻辑

genius-child-alice committed:
✨  统一字段命名并添加创建/编辑者信息

- 将 createDateTime 统一重命名为 createdDateTime
- 将 editDateTime 统一重命名为 editedDateTime
- 新增 createdBy 和 editedBy 字段记录操作者
- 更新相关索引和查询逻辑

genius-child-alice committed:
🐛 使用用户UID替代UUID生成会话ID避免暴露用户uuid

genius-child-alice committed:
🐛 移除重复的检查拉黑逻辑，使用BlockService中的方法

使用BlockService中已存在的checkIsBlockedByOtherUserService方法来替代ImService中重复的检查用户是否被拉黑的逻辑，保持代码一致性并减少维护成本

genius-child-alice committed:
🎨 将检查用户关注功能从ImService移到FeedService

将检查用户是否关注的功能从ImService迁移到FeedService中，并完善了参数验证和错误处理逻辑，使功能归属更加合理

genius-child-alice committed:
🎨 为多个函数添加JSDoc注释说明参数和返回值

genius-child-alice committed:
🐛 优化会话管理逻辑并修复参数命名

重构 getOrCreateConversation 方法，使用更清晰的参数命名 currentUserUuid 和 otherUserUid
修复发送消息时的自检逻辑，改为比较 UID 而非 UUID
确保会话创建和恢复时使用正确的用户标识

genius-child-alice committed:
🐛 修改未回复消息检查逻辑，允许发送3条消息

将未回复消息检查逻辑从"发送过消息"改为"发送过3条或更多消息"才触发限制
同时修改关注检查方向，改为检查接收者是否关注发送者

genius-child-alice committed:
✨ 添加消息审核的TODO注释

genius-child-alice committed:
⚡️ 为getOrCreateConversation添加MongoDB会话支持

支持在事务中使用getOrCreateConversation方法，通过可选参数传递MongoDB会话

genius-child-alice committed:
⚡️ 修复撤回消息时内容显示问题

撤回消息时不再显示"[消息已撤回]"而是直接返回空内容

genius-child-alice committed:
⚡️ 修复删除状态查询条件不一致问题

将删除状态的查询条件从`{ $ne: true }`统一改为`false`，保持代码一致性并避免潜在的类型问题

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+			// 检查接收者的最后一条消息是否在发送者的最后一条消息之后
+			const lastSenderMessage = senderMessages.result.sort((a, b) => b.createdDateTime - a.createdDateTime)[0]
+			const lastReceiverMessage = receiverMessages.result.sort((a, b) => b.createdDateTime - a.createdDateTime)[0]
+			return lastReceiverMessage.createdDateTime < lastSenderMessage.createdDateTime
```
这个地方读起来像是 A 只要发送了三条，那么 B 就必须回复一条，A 才能继续发送。
我觉得应该是 B 回复一条之后 A 就可以无限发送了。

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+					conversationId: 1,
+					otherUser: {
+						uid: '$otherUserInfo.uid',
+						uuid: '$otherUserUuid',
```
不暴露对方 UUID

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+					},
+					lastMessage: {
+						$cond: {
+							if: { $ne: ['$lastMessageData', null] },
```
返回已撤回和已删除标识，而不是字符串。因为前端会做国际化。而且逻辑判断一般不发给数据库执行。

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+				{ user1Uuid: uuid },
+				{ user2Uuid: uuid },
+			],
+		} as any
```
no-any

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+				{ user1Uuid: uuid },
+				{ user2Uuid: uuid },
+			],
+		} as any
```
no-any

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+				{ user1Uuid: uuid },
+				{ user2Uuid: uuid },
+			],
+		} as any
```
no-any

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+				{ senderUuid: uuid },
+				{ receiverUuid: uuid },
+			],
+		} as any
```
no-any

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1240 @@
+ * @param session 可选的 MongoDB 会话（用于事务）
+ * @returns 包含成功状态和会话信息的对象。如果成功，返回 { success: true, conversation: ... }；如果失败，返回 { success: false }
+ */
+const getOrCreateConversation = async (currentUserUuid: string, otherUserUid: number, session?: any): Promise<{ success: boolean; conversation?: InferSchemaType<typeof ImConversationSchema.schemaInstance> }> => {
```
no-any

genius-child-alice committed:
🐛 修改未回复消息检查逻辑以允许曾经回复过的对话

调整 checkHasUnrepliedMessage 函数的逻辑，现在只要接收者曾经回复过（无论时间顺序），就允许发送者继续发送消息，而不是比较最后一条消息的时间

genius-child-alice committed:
refactor(IM): 优化会话列表数据结构并添加消息状态字段

移除不再使用的uuid字段，添加isRecalled和isDeleted字段来跟踪消息状态
重构会话列表映射逻辑，根据发送者判断消息删除状态

genius-child-alice committed:
refactor(ImService): 改进类型安全并移除any类型

- 使用ClientSession替代any类型作为会话参数类型
- 将未知类型转换为具体类型Record<string, unknown>
- 移除不必要的any类型注解，使用更精确的类型定义

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1260 @@
+			return {
+				messageId: (itemData.messageId as string) || '',
+				senderUuid: (itemData.senderUuid as string) || '',
+				receiverUuid: (itemData.receiverUuid as string) || '',
```
检查不要返回 UUID

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1260 @@
+				$skip: skip,
+			},
+			{
+				$limit: pageSize,
```
通过 pagesize 的方式获取消息会有问题。

假设 pagesize 为 3，第一次进入时，获取消息 `{ page: 1, pageSize: 3 } -> { skip: 0, limit: 3 }`
```
CCC
BBB
AAA
```
用户向上滚动，触发重新获取消息 `{ page: 2, pageSize: 3 } -> { skip: 3, limit: 3 }`
```
FFF
EEE
DDD
```
新数据会置于原数据前，得到：
```
FFF
EEE
DDD
CCC
BBB
AAA
```
假设此时对方发送消息 `foo`，暂且先不提如何获取到这条新发送的 `foo` 这个问题。
如果用户仍然向上滚动，再次触发获取消息 `{ page: 3, pageSize: 3 } -> { skip: 6, limit: 3 }`
请求结果会变成：
```
HHH
GGG
FFF
```
该数据与原始数据拼接后，消息 `FFF` 会重复出现。且当对方在你滚动查看历史消息期间发送的新消息数量超过 pageSize 后，你再也无法获取历史消息。

—

一个解决思路是：不再永远以集合最前文档作为 skip 计数的起点，而是以用户开始向上滚动时的那一条消息，从"绝对定位"改为"相对定位"。至于有没有性能好的实现方法，需要仔细考虑和测试

cfdxkk reviewed (changes requested):
脑内运行了一下，发现了一个比较重要的业务逻辑问题。

cfdxkk commented on src/service/ImService.ts:
```diff
@@ -0,0 +1,1260 @@
+				{ user1Uuid: uuid },
+				{ user2Uuid: uuid },
+			],
+		} as QueryType<Conversation>
```
直接在常量定义时声明类型 `const foo: Foo = xxx`，而不是给对象字面量 as type 后初始化给常量。
`QueryType` 可能不支持 or，如果 or 是合法的，就需要去改一下这个自定义的类型，因为它不一定完善，只包含了我已经使用过的一下查询条件，不是全部支持的查询条件。

genius-child-alice committed:
🐛 返回消息的 senderUuid 和 receiverUuid 字段改为 senderUid 和 receiverUid

genius-child-alice committed:
🐛 实现基于游标的消息列表分页加载，防止随时增长的消息对加载历史消息造成重复或消失的问题

genius-child-alice committed:
🐛 在常量声明处标注类型，补全QueryType对$or的类型支持

genius-child-alice committed:
🎨  使用limitPageSize工具函数限制分页大小

genius-child-alice committed:
🐛 修复getMessageListService缺少$skip的bug

genius-child-alice committed:
🎨 修正代码格式

genius-child-alice committed:
🎨 修正代码格式

genius-child-alice committed:
✨ 添加私信隐私设置以及发送消息时的检查功能

genius-child-alice committed:
✨ 增加会话列表的关注状态筛选功能，允许筛选和我不同关注状态的用户之间的会话
